test/parse/test_entity_declaration.rb

```diff
@@ -23,6 +23,239 @@ def parse(internal_subset)
+          end
+        end
+
+        class TestExternalId < self
+          # https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-ExternalID
+          class TestSystemLiteral < self
+            def test_no_quote_in_system
+              # https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-SystemLiteral
+              exception = assert_raise(REXML::ParseException) do
+                REXML::Document.new('<!DOCTYPE root [<!ENTITY % valid-name SYSTEM invalid-system-literal > ]>')
+              end
+              assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 72
+Last 80 unconsumed characters:
+ % valid-name SYSTEM invalid-system-literal > ]>
+              DETAIL
+            end
+
+            def test_no_quote_in_public
+              # https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-SystemLiteral
+              exception = assert_raise(REXML::ParseException) do
+                REXML::Document.new('<!DOCTYPE root [<!ENTITY % valid-name PUBLIC "valid-pubid-literal" invalid-system-literal > ]>')
+              end
+              assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 94
+Last 80 unconsumed characters:
+ % valid-name PUBLIC \"valid-pubid-literal\" invalid-system-literal > ]>
+              DETAIL
+            end
+          end
+
+          class TestPubidLiteral < self
+            def test_no_quote
+              # https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-PubidLiteral
+              exception = assert_raise(REXML::ParseException) do
+                REXML::Document.new('<!DOCTYPE root [<!ENTITY % valid-name PUBLIC invalid-pubid-literal "valid-system-literal" > ]>')
+              end
+              assert_equal(<<-DETAIL.chomp, exception.to_s)
+Malformed entity declaration
+Line: 1
+Position: 94
+Last 80 unconsumed characters:
+ % valid-name PUBLIC invalid-pubid-literal \"valid-system-literal\" > ]>
+              DETAIL
+            end
+
+            def test_invalid_pubid_char
+              # https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-PubidChar
+              exception = assert_raise(REXML::ParseException) do
+                # U+3042 HIRAGANA LETTER A
+                REXML::Document.new("<!DOCTYPE root [<!ENTITY % valid-name PUBLIC \"\u3042\" \"valid-system-literal\" > ]>")
+              end
+              assert_equal(<<-DETAIL.force_encoding('utf-8').chomp, exception.to_s.force_encoding('utf-8'))
+Malformed entity declaration
+Line: 1
+Position: 78
+Last 80 unconsumed characters:
+ % valid-name PUBLIC \"\u3042\" \"valid-system-literal\" > ]>
+              DETAIL
+            end
+          end
+        end
+      end
+
+      def test_unnecessary_ndata_declaration
+        # https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-PEDef
+        exception = assert_raise(REXML::ParseException) do
+          REXML::Document.new('<!DOCTYPE root [<!ENTITY % valid-name "valid-entity-value" "NDATA" valid-ndata-value > ]>')
```

Comment:
```suggestion
          REXML::Document.new('<!DOCTYPE root [<!ENTITY % valid-name "valid-entity-value" NDATA valid-ndata-value > ]>')
```

BTW, should we use the `ExternalID NDataDecl?` part instead of the `EntityValue` part in `EntityDef	   ::= EntityValue | ExternalID NDataDecl?)`? If we want to use the `EntityValue` part, we should use "garbage" or something instead of `ndata_declaration` because all contents (including `NDataDecl`) are invalid after the `EntityValue`.

Reply:
Make sence.
We should make sure to check that any unnecessary NDataDecl after EntityValue will raise an exception.

Reply:
Added test at https://github.com/ruby/rexml/pull/183/commits/478208469566827458bb871ec132451c5f660cfa